### DISCUSSION TITLE
Decoding an arbitrary vector (passed as a transaction parameter)

### DISCUSSION THREAD

**Discussion by aeryz** [aptos-labs/aptos-developer-discussions#389] [Questions]
### Describe your question in detail.

Hey, I'm implementing IBC protocol in aptos-move. Some of the `entry` functions we have expects `proof` parameter which should be just arbitrary bytes as the protocol spec suggests. This `proof` is then decoded by the corresponding light client (the proof type might change based on the light client).

To do this, we tried to use `Any` because it's not allowed to decode bytes into arbitrary types. But then realized `Any` is not allowed as a transaction parameter. Then we tried having a function to manually decode arbitrary bytes using `bcs` but again realized that there is no support for partial decoding (peel) similar to [here](https://docs.sui.io/references/framework/sui-framework/bcs).

For now we are using our protobuf helpers to decode arbitrary types. But I would love to not do that and just be able to encode/decode as easy as possible. Is there a way to achieve this?

### Which operating system are you using?

Linux (Ubuntu, Fedora, Windows WSL, etc.)

### Which SDK or tool are you using? (if any)

N/A

**Comment by 0x-j** [comment]
do functions in this contract help you decode? https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-stdlib/sources/from_bcs.move

I'm not familiar with decoding/encoding, so please bear with me

**Reply by aeryz** [reply]
Hey! Thanks for the answer. Unfortunately this module does not export `from_bytes` function which lets you decode bytes into an arbitrary type: https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-stdlib/sources/from_bcs.move#L67.

I also tried having a function to manually decode the fields. But `from_bcs::to_xxx` functions don't do partial decoding. So it didn't work either.

**Reply by 0x-j** [reply]
hey i just dm'ed you on twitter, maybe we can have a meeting to chat this in detail

**Reply by aeryz** [reply]
I created an `bcs_utils` library to let people do partial decoding:
https://gist.github.com/aeryz/d005e7ccd2f12da1d0e58ed4bcf2a36d

And here is how you would use it for your type:
```
    public fun decode_existence_proof(buf: &mut BcsBuf): ExistenceProof {
        ExistenceProof {
            key: bcs_utils::peel_bytes(buf),
            value: bcs_utils::peel_bytes(buf),
            leaf_prefix: bcs_utils::peel_bytes(buf),
            path:bcs_utils::peel_vector<InnerOp>(buf, |buf| InnerOp {
                prefix: bcs_utils::peel_bytes(buf),
                suffix: bcs_utils::peel_bytes(buf),
            }),
        }
    }
```
